examples/getting_started/tools_and_agents.ipynb

```diff
@@ -0,0 +1,586 @@
+    "\n",
+    "In this notebook, we'll explore how to create and use Tools, and how to implement an Agent using the WebSearchAgent as our example. We'll be using Mirascope, a toolkit that simplifies the process of building LLM-powered applications with tools and agents.\n",
+    "\n",
+    "For more detailed information on these concepts, refer to the following Mirascope documentation:\n",
+    "\n",
+    "- [Tools documentation](https://docs.mirascope.io/latest/learn/tools/)\n",
+    "- [Agents documentation](https://docs.mirascope.io/latest/learn/agents/)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 2. Setting Up the Environment\n",
+    "\n",
+    "First, let's set up our environment by installing Mirascope and other necessary libraries."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install \"mirascope[openai]\" duckduckgo_search beautifulsoup4 requests"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "\n",
+    "os.environ[\"OPENAI_API_KEY\"] = \"your-api-key-here\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": "For more information on setting up Mirascope and its dependencies, see the [Mirascope installation guide](https://docs.mirascope.io/latest/get-started/)."
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Building a Basic Chatbot\n",
+    "\n",
+    "Before we dive into Tools and Agents, let's start by building a basic chatbot. This will help us understand the fundamental concepts of state management and conversation flow in Mirascope."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-09-12T01:39:45.815456Z",
+     "start_time": "2024-09-12T01:39:20.717191Z"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "User: Good morning!\n",
+      "Chatbot: Good morning! How are you today?\n",
+      "User: I'm doing well, thank you! How about you?\n",
+      "Chatbot: I’m doing great, thanks for asking! Is there anything specific you’d like to chat about today?\n"
+     ]
+    }
+   ],
+   "source": [
+    "from pydantic import BaseModel\n",
+    "\n",
+    "from mirascope.core import BaseMessageParam, openai, prompt_template\n",
+    "\n",
+    "\n",
+    "class BasicChatbot(BaseModel):\n",
+    "    messages: list[BaseMessageParam | openai.OpenAIMessageParam] = []\n",
+    "\n",
+    "    @openai.call(model=\"gpt-4o-mini\")\n",
+    "    @prompt_template(\n",
+    "        \"\"\"\n",
+    "        SYSTEM: You are a friendly chatbot assistant. Engage in a conversation with the user.\n",
+    "        \n",
```

Comment:
nit: remove remove newline here between the system message and the MESSAGES keyword